### PR TITLE
Ignore `unsopported constant expr` error

### LINKED ISSUE #46553: Incorrect constant evaluation warning on correct code
It looks like I'm incorrectly being warned of a constant evaluation error on valid code. The code runs just fine and is not violating the constant function requirements.

Offending code:
```rust
#![feature(const_fn)]

pub struct Data<T> {
    function: fn() -> T,
}

impl<T> Data<T> {
    pub const fn new(function: fn() -> T) -> Data<T> {
        Data {
            function: function,
        }
    }
}

pub static DATA: Data<i32> = Data::new(|| {
    413i32
});

fn main() {
    print!("{:?}", (DATA.function)());
}
```
Incorrect error:
```rust
warning: constant evaluation error: unsupported constant expr
  --> src/main.rs:15:30
   |
15 |   pub static DATA: Data<i32> = Data::new(|| {
   |  ______________________________^
16 | |     413i32
17 | | });
   | |__^
   |
   = note: #[warn(const_err)] on by default
```
[Example of the issue on the playground in nightly](https://play.rust-lang.org/?gist=201319d1da5c170986d22d096d493f0d&version=nightly).

Ping #24111

### PR DESCRIPTION
**Pull request by AgustinCB**
Fixes #46553

### CODE REVIEW

File: src/test/run-pass/issue-46553.rs [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+#![feature(const_fn)]
+#![warn(const_err)]
```
Comment (oli-obk) [inline src/test/run-pass/issue-46553.rs]: This should be deny, because otherwise the test won't fail if the warning is reintroduced

**Review by oli-obk — COMMENTED** [review]
You can write `fixes` instead of `addresses`, then github will automatically close the issue when the PR is merged

**Review by oli-obk — APPROVED** [review]

**Comment by mooman219** [comment]
From what I understand (stuff mentioned in IRC), the real issue is that rustc has two const evaluators. One for stuff like array lengths, and one for everything else. You are seeing a warning from the first one, but the second one processes this just fine. Is suppressing the error like this properly handling the issue?

**Comment by carols10cents** [comment]
review ping for you @pnkfelix !

**Comment by eddyb** [comment]
LGTM. @bors delegate=oli-obk

**Comment by oli-obk** [comment]
Yea suppressing this seems fine to me, since the lint is meant to report const eval failures early. Implementation deficiencies in the hir const evaluator could be fixed, but it would be wasted work with miri being so close.

@bors r+

**Pull request merged**